### PR TITLE
Add periodic test for containerd using PowerVS

### PR DESCRIPTION
**Pull request by EGuesnet**
This PR adds periodic test for containerd using PowerVS (full test, using different configuration, to mimic upstream CI).
Tested here: https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/periodic-containerd-test-powerVS/1504756840649461760 .

### CODE REVIEW

File: config/jobs/periodic/containerd/test-containerd-powervs-periodic.yaml
Comment (Rajalakshmi-Girish) [inline config/jobs/periodic/containerd/test-containerd-powervs-periodic.yaml:42]: From the build log of job https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/periodic-containerd-test-powerVS/1504756840649461760 I see tests for the below configuration does not seem to have run:

`--runc runc --runtime "io.containerd.runc.v2"`

The below process seems to be abruptly **missing/failing** but still, the prow job result is seen as PASSED:
```
bash -x ./instanciate_powervs_vm.sh --output ${ARTIFACTS} --name rdr-runtimes-dev-etienne_guesnet-crun_v2 --runc crun --runtime "io.containerd.runc.v2" --key atos-key --network 'public-192_168_141_8-29-VLAN_2093' &
```
Looks like `instanciate_powervs_vm.sh` needs improvisation to tell prow job if there is any failure while VM deployment.

Also, I see all the junit xml files from four different configurations being placed under the same ARTIFACTS folder

(I understand that they have different naming conventions) hence the Junit tab in [UI](https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/periodic-containerd-test-powerVS/1504756840649461760) does not distinguish the tests with respect to configuration under which it is run.

@EGuesnet you may explore how to make the UI distinguish based on the configuration in which it was run. 

@alunsin ^^

File: config/jobs/periodic/containerd/test-containerd-powervs-periodic.yaml
Comment (Rajalakshmi-Girish) [inline config/jobs/periodic/containerd/test-containerd-powervs-periodic.yaml:56]: I suggest having a timeout for this too.

**Review by Rajalakshmi-Girish — CHANGES_REQUESTED** [review]

**Comment by EGuesnet** [comment]
Hi @Rajalakshmi-Girish , 
This PR is on my top priority, but PowerVS fails a lot (half of VM requested fail...), so I cannot test my changes or correctly run test.

**Comment by Rajalakshmi-Girish** [comment]
> Hi @Rajalakshmi-Girish , This PR is on my top priority, but PowerVS fails a lot (half of VM requested fail...), so I cannot test my changes or correctly run test.

@EGuesnet True. I have been seeing a lot of job failures due to PowerVS failure to provision VM. This is observed in the `lon04` zone. Which region/zone are you trying with?
Our jobs get the below error when trying to provision VM in PowerVS:
```
31m│[0m [0m[1m[31mError: [0m[0m[1mfailed to provision: failed to Create PVM Instance :&{0   } (*models.Error) is not supported by the TextConsumer, can be resolved by supporting TextUnmarshaler interface[0m
```

**Comment by Rajalakshmi-Girish** [comment]
@mkumatag our conformance jobs are seeing intermittent failures with PowerVS since Friday evening. 

image

I see a similar support case raised by Manoj https://cloud.ibm.com/unifiedsupport/cases?number=CS2724230 for lpar creation failure.
The error message might be different due to TF versions.

**Comment by EGuesnet** [comment]
My region is London also (`ibmcloud target -r eu-gb -g runtimes-dev-resource-group`, runtimes-dev-lon06). Some colleagues use other locations (Brasil), and they experiment also some trouble.

**Comment by Rajalakshmi-Girish** [comment]
> My region is London also (`ibmcloud target -r eu-gb -g runtimes-dev-resource-group`, runtimes-dev-lon06). Some colleagues use other locations (Brasil), and they experiment also some trouble.

@EGuesnet we don't see any PowerVS VM creation failure these days. It seems to be pretty stable.
Were you able to test your changes?

**Comment by EGuesnet** [comment]
I test right now. On 4 VM, I have 2 failures:
- one stuck in grub
- the other one is correctly initialized, but the instance command returns a failure (probably because it takes too much time to respond), so tests are not run

I will try sequentially.

**Comment by EGuesnet** [comment]
Run sequentially, I have the grub trouble again.

**Comment by EGuesnet** [comment]
Hi @Rajalakshmi-Girish ,
I have removed useless variables.
If a VM cannot be instantiated, job result will be a fail. Junit files are in different subfolder.

In order to minimize the risk of PowerVS provisioning failure, I now run the test sequentially. It is longer, but it works better. Moreover, I have added lot of workaround to bypass various trouble. E.G. if the instantiation returns a failure, I even try to connect to the VM, because it works most of the time, and I return an error only if there is no VM accessible.

The updated prow job has worked two times without any error https://prow.ppc64le-cloud.org/job-history/s3/prow-logs/logs/periodic-containerd-test-powerVS .

**Comment by EGuesnet** [comment]
Hi @Rajalakshmi-Girish ,
Any news ? Is there any blocking point?

**Comment by Rajalakshmi-Girish** [comment]
> Hi @Rajalakshmi-Girish , Any news ? Is there any blocking point?

I shall review and get back today.

**Comment by Rajalakshmi-Girish** [comment]
> The updated prow job has worked two times without any error https://prow.ppc64le-cloud.org/job-history/s3/prow-logs/logs/periodic-containerd-test-powerVS .

In the run [1517272104402161664](https://s3.us-south.cloud-object-storage.appdomain.cloud/prow-logs/logs/periodic-containerd-test-powerVS/1517272104402161664/build-log.txt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=e8f95cb9812a478ebf9d2d6b89c75c85%2F20220504%2Fus-south%2Fs3%2Faws4_request&X-Amz-Date=20220504T035024Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=d84503b38a4d4152eef657f358d40f129e880a8aea9dd193399ab8b4c3250a99) I see an integration suite failure.
Please look for the text "FAIL	github.com/containerd/containerd/integration/client" to see this.

I feel we need a better way to look out for failures in the log. A simple grep for text "FAIL: " might not be work always.
Some failures where tests Panic due to goroutine leaks may not be caught this way.


Also, I thought of having a final run against our actual production ppc cluster before merging this change.
But our cluster is down now. I shall run this job once our cluster is fixed.

**Comment by EGuesnet** [comment]
The main goal of grepping "FAIL:" is to get troubles with machines. If a test fails, thanks to gotestsum, the job is correctly reported as failure. Unfortunately, integration-tests do not support gotestsum. This is known upstream, and it is in their improvement todolist.

We cannot grep "FAIL" only, because this keyword can occur in no-real fail conditions (PowerVS trouble we workaround, expected fail...).

In my opinion, the best way to track all error is to add gotestsum support upstream for integration-test.

**Comment by Rajalakshmi-Girish** [comment]
> Unfortunately, integration-tests do not support gotestsum.

The job yaml /lgtm
But, if the integration-tests don't throw failure with "FAIL:" text, then are you not missing the validation for this suite?

Is there a way you plan to catch this failure till the upstream improves this?

**Comment by Rajalakshmi-Girish** [comment]
> I will test it Monday and take you inform of the result

@EGuesnet Thanks for the detailed reply.

Please update once the changed regex and grep work fine in catching the cri-integration failures.

Is the configuration tested through job `periodic-containerd-test-ppc64le` also covered in one of the use cases in `periodic-containerd-test-powerVS `. Will the old job be removed once this new one gets added?

**Comment by EGuesnet** [comment]
@Rajalakshmi-Girish ,
I am not able to run all tests, PowerVS does not work smoothly: https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/periodic-containerd-test-powerVS/1524735660856971264 . The creation of the 4th machine has failed (to be precise, the VM is provisioned and can get a public IP after 5 or 6 minutes usually; the timeout has failed after 10 minutes).
I check the final regex on logs, and it works as expected. But it does not appear in the last run, because the job fails before the end.

The last patch only increases the timeout.

> Is the configuration tested through job periodic-containerd-test-ppc64le also covered in one of the use cases in periodic-containerd-test-powerVS . Will the old job be removed once this new one gets added?

``periodic-containerd-test-ppc64le`` is really more stable than ``periodic-containerd-test-powerVS`` because it does not use PowerVS. If it is not too resources consuming, I recommend to keep `periodic-containerd-test-ppc64le`.

**Comment by Rajalakshmi-Girish** [comment]
> But it does not appear in the last run, because the job fails before the end.

@EGuesnet I think it would be good to proceed only after a successful test run. Maybe we can wait till @trex58i takes this up and does a re-run when PowerVs is in better condition. We can put this on hold till then.

I see you running this in `lon06` zone. Our jobs on lon04 are decently good at VM creation and reach. You may try switching the region/zone on PowerVS.

@mkumatag ^^

**Comment by EGuesnet** [comment]
`lon04` has the same trouble. I put this PR on hold. @trex58i will take the next, when he will have time to do.

**Comment by EGuesnet** [comment]
@trex58i , when you will work on this PR, beware of names: I name VM something like `--name rdr-runtimes-dev-**etienne_guesnet**-runc-v1`. You will need to change that.
Name is required to know who is in charge of the VM (e.g. in case of security trouble).

**Comment by Rajalakshmi-Girish** [comment]
https://github.com/ppc64le-cloud/test-infra/pull/324 has been raised as a replacement for this PR.
Hence closing this.

**Pull request closed without merging**
